src/controllers/plugins/api/HTTP.hpp

```diff
@@ -0,0 +1,152 @@
+#pragma once
+#include "common/network/NetworkRequest.hpp"
+#include "common/network/NetworkResult.hpp"
+#include "controllers/plugins/LuaUtilities.hpp"
+#include "controllers/plugins/PluginController.hpp"
+
+#include <memory>
+#ifdef CHATTERINO_HAVE_PLUGINS
+
+namespace chatterino::lua::api {
+// NOLINTBEGIN(readability-identifier-naming)
+
+/**
+ * @lua@class HTTPResult
+ * @lua@field data string Data received from the server
+ * @lua@field status integer? HTTP Status code returned by the server
+ * @lua@field error string A somewhat human readable description of an error if such happened
+ */
+
+/**
+ * @lua@alias HTTPCallback fun(result: HTTPResult): nil
+ */
+
+/**
+ * @lua@class HTTPRequest
+ */
+class HTTPRequest : public std::enable_shared_from_this<HTTPRequest>
+{
+    // This type is private to prevent the accidental construction of HTTPRequest without a shared pointer
+    struct ConstructorAccessTag {
+    };
+
+public:
+    HTTPRequest(HTTPRequest::ConstructorAccessTag, NetworkRequest req);
+    HTTPRequest(HTTPRequest &&other) = default;
+    HTTPRequest &operator=(HTTPRequest &&) = default;
+    HTTPRequest &operator=(HTTPRequest &) = delete;
+    HTTPRequest(const HTTPRequest &other) = delete;
+    ~HTTPRequest();
+
+private:
+    NetworkRequest req_;
+
+    static void createMetatable(lua_State *L);
+    friend class chatterino::PluginController;
+
+    /**
+     * @brief Get the content of the top object on Lua stack, usually the first argument as an HTTPRequest
+     *
+     * If the object given is not a userdatum or the pointer inside that
+     * userdatum doesn't point to a HTTPRequest, a lua error is thrown.
+     */
+    static std::shared_ptr<HTTPRequest> getOrError(lua_State *L,
+                                                   StackIdx where = -1);
+    /**
+     * Pushes the private table onto the lua stack
+     */
+    StackIdx pushPrivate(lua_State *L);
+
+    QString privateKey;
+    int timeout_ = 10'000;
+    bool done = false;
+
+public:
+    /**
+     * Sets the success callback
+     *
+     * @lua@param callback HTTPCallback Function to call when the HTTP request succeeds
+     * @exposed HTTPRequest:on_success
+     */
+    static int on_success_wrap(lua_State *L);
+    int on_success(lua_State *L);
```

Comment:
nit: why are these named `on_success` instead of `onSuccess`?
and what do their return values indicate?

Reply:
These are named `on_success` because they are API functions. This is the function you call to set the success callback. The return value is how many return values are on the stack.